### ISSUE TITLE
attempt to call method 'dim' (a nil value)

### ISSUE THREAD

**Issue by XinNoil** [nhammerla/deepHAR#1]
Hi,Thanks for your contribution. 
But when I want to run your code as your instruction , a error occurred. The print message as follow.

```
[program started on Tue Apr 24 14:43:15 2018]	
[command line arguments]	
cpu	true	
seed	1234	
learningRate	0.02516758	
maxInNorm	2.283772707	
batchSize	64	
data	../../data/opportunity.h5	
layerSize	179	
testEach	false	
logdir	EXP_brnn	
gpu	0	
maxOutNorm	0	
carryOverProb	0.915735543	
patience	10	
sequenceLength	81	
maxEpoch	150	
minEpoch	20	
numLayers	1	
[----------------------]	
Training epoch: 1	
/home/noil/torch/install/bin/luajit: /home/noil/torch/install/share/lua/5.1/nn/Container.lua:67: 
In 2 module of nn.Sequential:
In 2 module of nn.Sequential:
/home/noil/torch/install/share/lua/5.1/nn/JoinTable.lua:27: attempt to call method 'dim' (a nil value)
stack traceback:
	/home/noil/torch/install/share/lua/5.1/nn/JoinTable.lua:27: in function </home/noil/torch/install/share/lua/5.1/nn/JoinTable.lua:21>
	[C]: in function 'xpcall'
	/home/noil/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	/home/noil/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'updateOutput'
	/home/noil/torch/install/share/lua/5.1/nn/Decorator.lua:10: in function </home/noil/torch/install/share/lua/5.1/nn/Decorator.lua:9>
	[C]: in function 'xpcall'
	/home/noil/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	/home/noil/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	main_brnn.lua:162: in function 'opfunc'
	/home/noil/torch/install/share/lua/5.1/optim/adagrad.lua:31: in function 'adagrad'
	main_brnn.lua:178: in function 'train'
	main_brnn.lua:273: in main chunk
	[C]: in function 'dofile'
	...noil/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x00405d50

WARNING: If you see a stack trace below, it doesn't point to the place where this error occurred. Please use only the one above.
stack traceback:
	[C]: in function 'error'
	/home/noil/torch/install/share/lua/5.1/nn/Container.lua:67: in function 'rethrowErrors'
	/home/noil/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	main_brnn.lua:162: in function 'opfunc'
	/home/noil/torch/install/share/lua/5.1/optim/adagrad.lua:31: in function 'adagrad'
	main_brnn.lua:178: in function 'train'
	main_brnn.lua:273: in main chunk
	[C]: in function 'dofile'
	...noil/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x00405d50
```

**Comment by XinNoil** [comment]
Another question, may I get your other models' usage except for bi-RNN.

**Comment by XinNoil** [comment]
I have resolved this problem, it may be caused by that input x and labels in train function should be a tensor not a table.

**Comment by Hangwei12358** [comment]
> I have resolved this problem, it may be caused by that input x and labels in train function should be a tensor not a table.

Hi,
I've encountered the same problem, can you be so nice to share with me the solution?? Thanks!

**Comment by XinNoil** [comment]
> > I have resolved this problem, it may be caused by that input x and labels in train function should be a tensor not a table.
> 
> Hi,
> I've encountered the same problem, can you be so nice to share with me the solution?? Thanks!

My solution is here, modify the `dataIter` and `linearDataIter` function in `models/RNN/explib.lua`:
for example,the origin function is as follow:
```lua
-- iterator through the data-set (one particle per batch)
dataIter = function(data, labels)
    --local pos = 1
    -- random start positions for a particle for each batch
    local posList = torch.rand(params.batchSize)*(data:size(1)-params.sequenceLength)
    posList:ceil()
    posList = posList:long() -- need LongTensor for addressing

    local nelem = params.batchSize * params.sequenceLength

    local d
    local l
    return function()
        d = {}
        l = {}

        -- insert inputs into table
        for i=1,params.sequenceLength do
          table.insert(d, data:index(1, posList))
          table.insert(l, labels:index(1, posList))
          posList:add(1)
        end

        -- check if we have to reset positions
        for i=1,params.batchSize do
          if posList[i] + 1 > data:size(1)-params.sequenceLength then
            posList[i] = 1
          end
        end

        return d,l
    end
end
```
I modified it like this:
```lua
-- iterator through the data-set (one particle per batch)
dataIter = function(data, labels)
    --local pos = 1
    -- random start positions for a particle for each batch
    local posList = torch.rand(params.batchSize)*(data:size(1)-params.sequenceLength)
    posList:ceil()
    posList = posList:long() -- need LongTensor for addressing

    local nelem = params.batchSize * params.sequenceLength
    local d
    local l
    return function()
        d = torch.Tensor(params.sequenceLength,params.batchSize,data:size(2))
	l = torch.Tensor(params.sequenceLength,params.batchSize)
	-- d = {}
        -- l = {}
        -- print("posList:",posList)
        x=data:index(1, posList)
        -- print("x size:",#x," x type:",type(x)," x tType:",torch.type(x))
        -- insert inputs into table
        for i=1,params.sequenceLength do
            
	        d[{i,{},{}}] = data:index(1, posList)
	        l[{i,{}}] = labels:index(1,posList)
            -- table.insert(d, data:index(1, posList))
            -- table.insert(l, labels:index(1, posList))
            posList:add(1)
        end

        -- check if we have to reset positions
        for i=1,params.batchSize do
          if posList[i] + 1 > data:size(1)-params.sequenceLength then
            posList[i] = 1
          end
        end
        return d,l
    end
end
```

**Comment by Hangwei12358** [comment]
@XinNoil Thanks a lot for your quick reply!! 
I came across a similar solution with d = torch.cat(d, data:index(1, posList), 1) in the loop, and the code is still running. Hope I have refreshed the page earlier so that I can save many hours debugging...LOL...

**Comment by Hangwei12358** [comment]
@XinNoil Thanks a lot for your quick reply!!
I came across a similar solution with `d = torch.cat(d, data:index(1, posList), 1)` in the loop, and the code is still running. Hope I have refreshed the page earlier so that I can save many hours debugging...LOL...

**Comment by XinNoil** [comment]
> @XinNoil Thanks a lot for your quick reply!!
> I came across a similar solution with `d = torch.cat(d, data:index(1, posList), 1)` in the loop, and the code is still running. Hope I have refreshed the page earlier so that I can save many hours debugging...LOL...

It's really hard,I know... when I'm doing this, I even haven't studied lua.
Now I'm using keras for my project, it's more friendly for me, a beginner.

**Comment by Hangwei12358** [comment]
> > @XinNoil Thanks a lot for your quick reply!!
> > I came across a similar solution with `d = torch.cat(d, data:index(1, posList), 1)` in the loop, and the code is still running. Hope I have refreshed the page earlier so that I can save many hours debugging...LOL...
> 
> It's really hard,I know... when I'm doing this, I even haven't studied lua.
> Now I'm using keras for my project, it's more friendly for me, a beginner.

Same here... It's my first time debugging lua codes too... Are you still working on activity recognition related projects now?

**Comment by XinNoil** [comment]
> > > @XinNoil Thanks a lot for your quick reply!!
> > > I came across a similar solution with `d = torch.cat(d, data:index(1, posList), 1)` in the loop, and the code is still running. Hope I have refreshed the page earlier so that I can save many hours debugging...LOL...
> > 
> > 
> > It's really hard,I know... when I'm doing this, I even haven't studied lua.
> > Now I'm using keras for my project, it's more friendly for me, a beginner.
> 
> Same here... It's my first time debugging lua codes too... Are you still working on activity recognition related projects now?

I'm working on Indoor localization and navigation, also use some sensors like activity recognition, they are pretty connected.

**Comment by Shiv3901** [comment]
@XinNoil

Could you please provide the code for LinearDataIter as well, I am very new to lua and cannot figure it out myself.
